lib/logstash/inputs/s3.rb

```diff
@@ -134,13 +136,116 @@ def backup_to_dir(filename)
     end
   end
 
+  public
+  def process_files(queue)
+    objects = list_new_files
+
+    objects.each do |key|
+      @logger.debug("S3 input processing", :bucket => @bucket, :key => key)
+
+      lastmod = @s3bucket.objects[key].last_modified
+
+      process_log(queue, key)
+
+      sincedb.write(lastmod)
+    end
+  end # def process_files
+
+
   private
   def process_local_log(queue, filename)
-    @codec.decode(File.open(filename, 'rb')) do |event|
-      decorate(event)
-      queue << event
+    @logger.debug('Processing file', :filename => filename)
+
+    metadata = {}
+    # Currently codecs operates on bytes instead of stream.
+    # So all IO stuff: decompression, reading need to be done in the actual
+    # input and send as bytes to the codecs.
+    read_file(filename) do |line|
+      @codec.decode(line) do |event|
+        # We are making an assumption concerning cloudfront
+        # log format, the user will use the plain or the line codec
+        # and the message key will represent the actual line content.
+        # If the event is only metadata the event will be drop.
+        # This was the behavior of the pre 1.5 plugin.
+        #
+        # The line need to go through the codecs to replace
+        # unknown bytes in the log stream before doing a regexp match or
+        # you will get a `Error: invalid byte sequence in UTF-8'
+        if event_is_metadata?(event)
+          @logger.debug('Event is metadata, updating the current cloudfront metadata', :event => event)
+          update_metadata(metadata, event)
+        else
+          decorate(event)
+
+          event["cloudfront_version"] = metadata[:cloudfront_version] unless metadata[:cloudfront_version].nil?
```

Comment:
I would avoid as much as possible the usage of unless, is more clean when you use if not. But this is a personal style test.